tsl/src/compression/compression.c

```diff
@@ -2005,82 +2130,120 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
 				 errmsg("inserting into compressed chunk with unique constraints disabled"),
 				 errhint("Set timescaledb.enable_dml_decompression to TRUE.")));
 
-	Chunk *comp = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
-	Relation in_rel = relation_open(comp->table_id, RowExclusiveLock);
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
+	Relation in_rel = relation_open(compressed_chunk->table_id, RowExclusiveLock);
 
 	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
+	/* fetch index details from compressed chunk */
+	index_oid = get_index_on_compressed_chunk(uncompressed_chunk, compressed_chunk);
+
 	Bitmapset *key_columns = RelationGetIndexAttrBitmap(out_rel, INDEX_ATTR_BITMAP_KEY);
 	Bitmapset *null_columns = NULL;
 
-	int num_scankeys;
-	ScanKeyData *scankeys = build_scankeys(chunk->fd.hypertable_id,
-										   chunk->hypertable_relid,
-										   decompressor,
-										   key_columns,
-										   &null_columns,
-										   slot,
-										   &num_scankeys);
-
-	bms_free(key_columns);
-
 	/*
-	 * Using latest snapshot to scan the heap since we are doing this to build
-	 * the index on the uncompressed chunks in order to do speculative insertion
-	 * which is always built from all tuples (even in higher levels of isolation).
+	 * By default there is always an index on compressed chunk which has segmentby
+	 * columns. Use this index to fetch the required compress tuples to check for
+	 * unqiue constraint violation.
 	 */
-	TableScanDesc heapScan = table_beginscan(in_rel, GetLatestSnapshot(), num_scankeys, scankeys);
-
-	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
-		 compressed_tuple != NULL;
-		 compressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	if (OidIsValid(index_oid))
+	{
+		idxrel = index_open(index_oid, ExclusiveLock);
+		scankeys = fetch_index_scankeys(idxrel,
+										uncompressed_chunk->hypertable_relid,
+										slot,
+										&num_scankeys,
+										SEGMENTBY_KEYS);
+		index_scan = index_beginscan(in_rel, idxrel, GetLatestSnapshot(), num_scankeys, 0);
+		index_rescan(index_scan, scankeys, num_scankeys, NULL, 0);
+		index_slot = table_slot_create(in_rel, NULL);
+		/* fetch matching compressed tuple based on segmentby column values */
+		while (index_getnext_slot(index_scan, ForwardScanDirection, index_slot))
+		{
+			HeapTuple compressed_tuple = ExecFetchSlotHeapTuple(index_slot, false, NULL);
+			heap_deform_tuple(compressed_tuple,
+							  decompressor.in_desc,
+							  decompressor.compressed_datums,
+							  decompressor.compressed_is_nulls);
+			write_logical_replication_msg_decompression_start();
+			row_decompressor_decompress_row(&decompressor, NULL);
+			write_logical_replication_msg_decompression_end();
+
+			TM_FailureData tmfd;
+			TM_Result result pg_attribute_unused();
```

Comment:
This code was already present in the old code. Could you please add a comment to explain why it is safe to ignore the result of the operation?

Reply:
It seems you added also an `assert`. My question was more _why_ is it safe to ignore the result of the operation? Maybe you could add this information to the comment.